jscomp/syntax/src/res_core.ml

```diff
@@ -3920,36 +3916,56 @@ and parseListExpr ~startPos p =
             loc))
       [(Asttypes.Nolabel, Ast_helper.Exp.array ~loc listExprs)]
 
-(* Overparse ... and give a nice error message *)
-and parseNonSpreadExp ~msg p =
-  let () =
-    match p.Parser.token with
-    | DotDotDot ->
-      Parser.err p (Diagnostics.message msg);
-      Parser.next p
-    | _ -> ()
-  in
-  match p.Parser.token with
-  | token when Grammar.isExprStart token -> (
-    let expr = parseExpr p in
-    match p.Parser.token with
-    | Colon ->
-      Parser.next p;
-      let typ = parseTypExpr p in
-      let loc = mkLoc expr.pexp_loc.loc_start typ.ptyp_loc.loc_end in
-      Some (Ast_helper.Exp.constraint_ ~loc expr typ)
-    | _ -> Some expr)
-  | _ -> None
-
 and parseArrayExp p =
   let startPos = p.Parser.startPos in
   Parser.expect Lbracket p;
-  let exprs =
-    parseCommaDelimitedRegion p ~grammar:Grammar.ExprList ~closing:Rbracket
-      ~f:(parseNonSpreadExp ~msg:ErrorMessages.arrayExprSpread)
+  let split_by_spread exprs =
+    List.fold_left
+      (fun acc curr ->
+        match (curr, acc) with
+        | (true, expr, startPos, endPos), _ ->
+          (* find a spread expression, prepend a new sublist *)
+          ([], Some expr, startPos, endPos) :: acc
+        | ( (false, expr, startPos, _endPos),
+            (no_spreads, spread, _accStartPos, accEndPos) :: acc ) ->
+          (* find a non-spread expression, and the accumulated is not empty,
+           * prepend to the first sublist, and update the loc of the first sublist *)
+          (expr :: no_spreads, spread, startPos, accEndPos) :: acc
+        | (false, expr, startPos, endPos), [] ->
+          (* find a non-spread expression, and the accumulated is empty *)
+          [([expr], None, startPos, endPos)])
+      [] exprs
+  in
+  let listExprsRev =
+    parseCommaDelimitedReversedList p ~grammar:Grammar.ExprList ~closing:Rbracket
+      ~f:parseSpreadExprRegionWithLoc
   in
   Parser.expect Rbracket p;
-  Ast_helper.Exp.array ~loc:(mkLoc startPos p.prevEndPos) exprs
+  let loc = mkLoc startPos p.prevEndPos in
+  let collectExprs = function 
+    | [], Some spread, _startPos, _endPos -> [spread]
+    | exprs, Some spread, _startPos, _endPos -> (
```

Comment:
One thing I wasn't so sure if whether I should use `startPos` and `endPos` as the location.